flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/CommonMapInputTypeStrategy.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.types.inference.strategies;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.functions.FunctionDefinition;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.inference.ArgumentCount;
+import org.apache.flink.table.types.inference.CallContext;
+import org.apache.flink.table.types.inference.InputTypeStrategy;
+import org.apache.flink.table.types.inference.Signature;
+import org.apache.flink.table.types.inference.Signature.Argument;
+import org.apache.flink.table.types.logical.LegacyTypeInformationType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.LogicalTypeRoot;
+import org.apache.flink.table.types.logical.utils.LogicalTypeMerging;
+import org.apache.flink.table.types.utils.TypeConversions;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/** An {@link InputTypeStrategy} that expects that all arguments have a common map type. */
+@Internal
+public final class CommonMapInputTypeStrategy implements InputTypeStrategy {
```

Comment:
Can we create a `CommonCollectionInputTypeStrategy` class with this logic
and 2 subclasses `CommonMapInputTypeStrategy` and `CommonArrayInputTypeStrategy` which customize it a bit.

I'm asking since `CommonMapInputTypeStrategy` and `CommonArrayInputTypeStrategy`  are almost identical

Reply:
remove CommonArrayInputTypeStrategy may affect other code?

Reply:
it could be implemented in different ways, the main idea is to extract common logic into one place and reuse it

Reply:
ok, I will try it, by the way, I have a question about pyflink function defintion, the ci test said
```
    @property
 1562:   def map_union(self, map2) -> 'Expression':
        """
        Returns a map created by merging two maps, 'map1' and 'map2'. These two maps should have a
        common map type. If there are overlapping keys, the value from 'map2' will overwrite the
        value from 'map1'. If any of maps is null, return null.
        .. seealso:: :py:attr:`~Expression.map_union`
        """
        return _binary_op("mapUnion")(self, map2)

```
` pyflink/table/expression.py:1562: error: Too many arguments`
do you know the reason?

Reply:
ok, change it.

Reply:
> Can we create a `CommonCollectionInputTypeStrategy` class with this logic and 2 subclasses `CommonMapInputTypeStrategy` and `CommonArrayInputTypeStrategy` which customize it a bit?
> 
> I'm asking since `CommonMapInputTypeStrategy` and `CommonArrayInputTypeStrategy` are almost identical

@snuyanzin I have already achieved it. Do you think the change is ok?